File: cpp/problems/leetcode/0033_search_in_rotated_sorted_array/solution.cpp
// LeetCode 33
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l = 0, r = static_cast<int>(nums.size()) - 1;
        while (l <= r) {
            int m = (l + r) / 2;
            if (nums[m] == target) return m;
            if (nums[l] <= nums[m]) {
                if (nums[l] <= target && target < nums[m])
                    r = m - 1;
                else
                    l = m + 1;
            } else {
                if (nums[m] < target && target <= nums[r])
                    l = m + 1;
                else
                    r = m - 1;
            }
        }
        return -1;
    }
};

int main() {
    vector<int> a{4, 5, 6, 7, 0, 1, 2};
    assert(Solution().search(a, 0) == 4);
    assert(Solution().search(a, 3) == -1);
    vector<int> b{1};
    assert(Solution().search(b, 0) == -1);
    assert(Solution().search(b, 1) == 0);
    cout << "0033 OK" << endl;
    return 0;
}

File: python/problems/leetcode/0033_search_in_rotated_sorted_array/solution.py
"""LeetCode 33: Search in Rotated Sorted Array."""

from __future__ import annotations


class Solution:
    def search(self, nums: list[int], target: int) -> int:
        l, r = 0, len(nums) - 1
        while l <= r:
            m = (l + r) // 2
            if nums[m] == target:
                return m
            if nums[l] <= nums[m]:
                if nums[l] <= target < nums[m]:
                    r = m - 1
                else:
                    l = m + 1
            else:
                if nums[m] < target <= nums[r]:
                    l = m + 1
                else:
                    r = m - 1
        return -1


if __name__ == "__main__":
    assert Solution().search([4, 5, 6, 7, 0, 1, 2], 0) == 4
    assert Solution().search([4, 5, 6, 7, 0, 1, 2], 3) == -1
    assert Solution().search([1], 0) == -1
    assert Solution().search([1], 1) == 0
    print("0033 OK")
